app/gui/src/presenter/graph/widgets.rs

```diff
@@ -0,0 +1,352 @@
+//! Widgets controller.
+//!
+//! The Widgets Controller is responsible for querying the language server for  information about
+//! the node's widget metadata or resolving from local cache.
+
+use crate::prelude::*;
+
+use crate::executor::global::spawn_stream_handler;
+use crate::presenter::graph::visualization::manager::Manager;
+use crate::presenter::graph::visualization::manager::Notification;
+use crate::presenter::graph::ViewNodeId;
+
+use controller::ExecutedGraph;
+use engine_protocol::language_server::SuggestionId;
+use ensogl::define_endpoints_2;
+use ide_view::graph_editor::component::node::input::widget;
+use ide_view::graph_editor::component::visualization;
+use ide_view::graph_editor::component::visualization::Metadata;
+use ide_view::graph_editor::data::enso::Code;
+use ide_view::graph_editor::WidgetUpdate;
+use ide_view::graph_editor::WidgetUpdates;
+
+
+
+/// =================
+/// === Constants ===
+/// =================
+
+const WIDGET_VISUALIZATION_MODULE: &str = "Standard.Visualization.Widgets";
+const WIDGET_VISUALIZATION_METHOD: &str = "get_full_annotations_json";
+
+
+
+// ===============
+// === Widgets ===
+// ===============
+
+define_endpoints_2! {
+    Input {
+        /// Create or update widget query with given definition.
+        request_widgets(Request),
+        /// Remove all widget queries of given node that are not on this list.
+        retain_node_expressions(ViewNodeId, HashSet<ast::Id>),
+        /// Remove all widget queries of given node.
+        remove_node(ViewNodeId),
+    }
+    Output {
+        /// Emitted when the node's visualization has been set.
+        widget_data(ViewNodeId, WidgetUpdates),
+    }
+}
+
+/// Graph widgets controller. One requester per function call. Responsible for querying the language
+/// server for information about the node's widget metadata or resolving from local cache.
+#[derive(Debug)]
+pub struct Widgets {
```

Comment:
This should be named `Controller` and should be used as `widgets::Controller`. Now the name is misleading, as this struct does not contain widgets.